include/ultra64/rcp.h

```diff
@@ -1,139 +1,827 @@
+#define RDRAM_DEVICE_ID_REG     (RDRAM_BASE_REG + 0x04)
+#define RDRAM_DELAY_REG         (RDRAM_BASE_REG + 0x08)
+#define RDRAM_MODE_REG          (RDRAM_BASE_REG + 0x0c)
+#define RDRAM_REF_INTERVAL_REG  (RDRAM_BASE_REG + 0x10)
+#define RDRAM_REF_ROW_REG       (RDRAM_BASE_REG + 0x14)
+#define RDRAM_RAS_INTERVAL_REG  (RDRAM_BASE_REG + 0x18)
+#define RDRAM_MIN_INTERVAL_REG  (RDRAM_BASE_REG + 0x1c)
+#define RDRAM_ADDR_SELECT_REG   (RDRAM_BASE_REG + 0x20)
+#define RDRAM_DEVICE_MANUF_REG  (RDRAM_BASE_REG + 0x24)
+
+#define RDRAM_0_DEVICE_ID   0
+#define RDRAM_1_DEVICE_ID   1
+
+#define RDRAM_RESET_MODE    0
+#define RDRAM_ACTIVE_MODE   1
+#define RDRAM_STANDBY_MODE  2
+
+#define RDRAM_LENGTH            (2 * 512 * 2048)
+#define RDRAM_0_BASE_ADDRESS    (RDRAM_0_DEVICE_ID * RDRAM_LENGTH)
+#define RDRAM_1_BASE_ADDRESS    (RDRAM_1_DEVICE_ID * RDRAM_LENGTH)
+
+#define RDRAM_0_CONFIG      0x00000
+#define RDRAM_1_CONFIG      0x00400
+#define RDRAM_GLOBAL_CONFIG 0x80000
+
+/**
+ * PIF Physical memory map (total size = 2 KB)
+ *
+ *            Size      Description     Mode
+ *  1FC007FF    +-------+-----------------+-----+
+ *              |  64 B | JoyChannel RAM  | R/W |
+ *  1FC007C0    +-------+-----------------+-----+
+ *              |1984 B |    Boot ROM     |  *  |  * = Reserved
+ *  1FC00000    +-------+-----------------+-----+
+ */
+#define PIF_ROM_START   0x1FC00000
+#define PIF_ROM_END     0x1FC007BF
+#define PIF_RAM_START   0x1FC007C0
+#define PIF_RAM_END     0x1FC007FF
+
+
+/**
+ * Controller channel 
+ * Each game controller channel has 4 error bits that are defined in bit 6-7 of
+ * the Rx and Tx data size area bytes. Programmers need to clear these bits
+ * when setting the Tx/Rx size area values for a channel
+ */
+#define CHNL_ERR_NORESP		0x80	/* Bit 7 (Rx): No response error */
+#define CHNL_ERR_OVERRUN	0x40	/* Bit 6 (Rx): Overrun error */
+#define CHNL_ERR_FRAME		0x80	/* Bit 7 (Tx): Frame error */
+#define CHNL_ERR_COLLISION	0x40	/* Bit 6 (Tx): Collision error */
+
+#define CHNL_ERR_MASK		0xC0	/* Bit 6-7: channel errors */
+
+
+/**
+ * External device info
+ */
+#define DEVICE_TYPE_CART	0	/* ROM cartridge */
+#define DEVICE_TYPE_BULK	1	/* ROM bulk */
+#define DEVICE_TYPE_64DD	2	/* 64 Disk Drive */
+#define DEVICE_TYPE_SRAM	3	/* SRAM */
+/* 4-6 are reserved */
+#define DEVICE_TYPE_INIT	7	/* initial value */
+/* 8-14 are reserved */
+
+/**
+ * Signal Processor (SP) Memory
+ */
+#define SP_DMEM_START   0x04000000
+#define SP_DMEM_END     0x04000FFF
+#define SP_IMEM_START   0x04001000
+#define SP_IMEM_END     0x04001FFF
+
+/**
+ * Signal Processor (SP) CP0 Registers
+ */
+
+#define SP_BASE_REG         0x04040000
+
+// SP memory address (R/W): [11:0] DMEM/IMEM address, [12] 0=DMEM,1=IMEM
+#define SP_MEM_ADDR_REG     (SP_BASE_REG + 0x00)
+
+// SP DRAM DMA address (R/W): [23:0] RDRAM address
+#define SP_DRAM_ADDR_REG    (SP_BASE_REG + 0x04)
+
+// SP read DMA length (R/W): [11:0] length, [19:12] count, [31:20] skip; RDRAM -> I/DMEM
```

Comment:
I don't like the bit ranges being in the opposite direction to the layout of the arguments. Since it's big-endian, to me at least it would make more sense to put [31:20] first etc, like the MIPS ISA document does in the instruction information, for example.